### PR TITLE
FSDP oneshot

### PR DESCRIPTION
**Pull request by Satrat**
This PR updates the one-shot modifiers SparseGPT, Wanda and SmoothQuant and Quantization to be compatible with FSDP. This enables us to run alternating one-shot/finetuning flows with FSDP

** NOTE: ** https://github.com/neuralmagic/sparseml/pull/1912 should be merged first, it covers the initial alternating flow implementation.

**Summary of Changes**
* ~~Remove any references of specific devices from the one-shot modifiers, device is now handled by `SparseCausalLM`, and defaults to "auto" for splitting the model across multiple GPUs (this isn't FSDP related, we can split the model even outside of FSDP)~~ "auto" actually isn't compatible with quantization :( so keeping the default as "cuda:0", but you can pass "auto" through the CLI for a non-quantized oneshot
* Refactored the `SparseGPT` class to be a module wrapper, so that we can update weights using module.apply as required for FSDP compatibility
* Refactored Wanda in the same way, also cleaned up the code sharing between SparseGPT and Wanda(@rahul-tuli would like your input specifically on this)
* Bug fixes related to quantizing FSDP models

### CODE REVIEW

File: src/sparseml/modifiers/obcq/utils/sgpt_wrapper.py [hunk 1]
```diff
@@ -183,28 +185,18 @@ def fasterprune(
...
+        W = W.reshape(final_shape).to(final_dtype)
+        self.layer.weight -= self.layer.weight
+        self.layer.weight += W
 
     def free(self):
         """
         Free the Hessian memory after the layer is complete
         """
-        self.H = None
+        delattr(self, "H")
```
Comment (rahul-tuli) [inline src/sparseml/modifiers/obcq/utils/sgpt_wrapper.py:201]: nice

File: tests/sparseml/core/recipe/test_stage.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                ignore: ["ReLU", "input"]
+    second_train_stage:
+        some_modifiers:
+            ConstantPruningModifier:
+                start: 0.0
+    """
+
+    recipe = Recipe.create_instance(recipe_str)
+    assert recipe.stages[0].infer_run_type() == StageRunType.ONESHOT
+    assert recipe.stages[1].infer_run_type() == StageRunType.TRAIN
```
Comment (rahul-tuli) [inline tests/sparseml/core/recipe/test_stage.py:52]: nice!

File: src/sparseml/modifiers/utils/compression_wrapper.py [hunk 3]
```diff
@@ -24,16 +26,13 @@
...
-DEBUG = False
-
 torch.backends.cuda.matmul.allow_tf32 = False
 torch.backends.cudnn.allow_tf32 = False
 
+__all__ = ["ModuleCompressionWrapper"]
+
 
-class ModuleCompressor(ABC):
+class ModuleCompressionWrapper(Module, ABC):
```
Comment (rahul-tuli) [inline src/sparseml/modifiers/utils/compression_wrapper.py:35]: 👍🏽

**Review by rahul-tuli — APPROVED** [review]
I like the new design a lot more, thanks for the good work here!

File: src/sparseml/modifiers/obcq/pytorch.py [hunk 4]
```diff
@@ -12,24 +12,26 @@
...
 
+_LOGGER = logging.getLogger(__name__)
+
 
 class SparseGPTModifierPyTorch(WandaPruningModifierPyTorch, SparseGPTModifier):
     """
     Pytorch implementation of SparseGPT
 
-    Lifecycle:
+    Lifecycle: TODO update
```
Comment (bfineran) [inline src/sparseml/modifiers/obcq/pytorch.py:34]: bump

File: src/sparseml/modifiers/obcq/utils/sgpt_wrapper.py [hunk 5]
```diff
@@ -183,28 +185,18 @@ def fasterprune(
...
 
         if isinstance(self.layer, transformers.Conv1D):
             W = W.t()
-        self.layer.weight.data = W.reshape(self.layer.weight.shape).to(
-            self.layer.weight.data.dtype
-        )
-        if DEBUG:
-            _LOGGER.debug(torch.sum((self.layer(self._inp1) - self.out1) ** 2))
+        W = W.reshape(final_shape).to(final_dtype)
+        self.layer.weight -= self.layer.weight
```
Comment (bfineran) [inline src/sparseml/modifiers/obcq/utils/sgpt_wrapper.py:194]: quick in-line comment

File: src/sparseml/modifiers/quantization/pytorch.py [hunk 6]
```diff
@@ -131,26 +129,29 @@ def _enable_module_qat(self, module: Module):
...
-            set_quantization_schemes(
-                module,
-                scheme=self.scheme,
-                scheme_overrides=self.scheme_overrides,
-                ignore=self.ignore,
-                strict=self.strict,
-            )
+            from torch.distributed.fsdp import FullyShardedDataParallel
+
+            with FullyShardedDataParallel.summon_full_params(module):
```
Comment (bfineran) [inline src/sparseml/modifiers/quantization/pytorch.py:134]: need to make this safe for pytorch < 2

maybe have try/except to assign a context manager with except as nullcontext

File: src/sparseml/modifiers/utils/layer_compressor.py [hunk 7]
```diff
@@ -12,73 +12,57 @@
...
 
-import inspect
 import logging
-from abc import ABC
-from typing import Dict, List
+import operator
+from typing import Dict
 
 import torch
+from torch.distributed.fsdp import FullyShardedDataParallel
```
Comment (bfineran) [inline src/sparseml/modifiers/utils/layer_compressor.py:20]: fdsp import again

File: src/sparseml/modifiers/utils/layer_compressor.py [hunk 8]
```diff
@@ -98,113 +82,48 @@ def pre_compress_parallel(self, **kwargs) -> Dict:
         """
         subset = self.compressible_modules()
 
-        gpts = {}
         for name in subset:
-            gpts[name] = self.module_compressor_class(subset[name])
+            layer = subset[name]
+            with FullyShardedDataParallel.summon_full_params(self.layer):
```
Comment (bfineran) [inline src/sparseml/modifiers/utils/layer_compressor.py:87]: maybe just make an optional helper function for this

File: src/sparseml/transformers/finetune/runner.py [hunk 9]
```diff
@@ -13,21 +13,28 @@
 # limitations under the License.
 
 import logging
-from typing import List
+import os
+from typing import List, Optional
 
 import torch
+from torch.distributed.fsdp import (
```
Comment (bfineran) [inline src/sparseml/transformers/finetune/runner.py:20]: fsdp import again

**Review by bfineran — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

File: src/sparseml/transformers/sparsification/obcq/example.yaml [hunk 10]
```diff
@@ -25,7 +25,7 @@ test_stage:
     SparseGPTModifier:
       sparsity: 0.5
       block_size: 128
-      sequential_update: False
+      sequential_update: True
```
[the remark this answers is not in the document]
Comment (Satrat) [reply src/sparseml/transformers/sparsification/obcq/example.yaml]: Whoops, yes it is. This was leftover from testing, just pushed a revert

**Comment by Satrat** [comment]
> > Remove any references of specific devices from the one-shot modifiers, device is now handled by SparseCausalLM, and defaults to "auto" for splitting the model across multiple GPUs (this isn't FSDP related, we can split the model even outside of FSDP)
> 
> It doesn't seem like device defaults to "auto" if that is an intended change. Current obcq.py arg:
> 
> ```
>     parser.add_argument("--device", type=str, default="cuda:0")
> ```

See updated PR comment :( device_map="auto" doesn't seem to be compatible with quantization so I'm leaving it off the default. It can still be specified on the CLI for non-quantized one-shot

**Review by bfineran — DISMISSED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**